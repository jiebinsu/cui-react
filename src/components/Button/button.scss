/* Buttons  */
@import '@coopdigital/foundations-buttons/dist/buttons.css';