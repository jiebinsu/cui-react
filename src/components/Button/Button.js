import React from "react";
import PropTypes from "prop-types";
import classNames from "../../utils/classNames";
import "./button.scss";

const Button = ({
  className,
  variant,
  type,
  size,
  disabled,
  fullWidth,
  isLoading,
  children,
  ...props
}) => {
  const dynamicClasses = classNames("coop-btn", [
    variant && `coop-btn--${variant}`,
    fullWidth && "coop-btn--full-width",
    isLoading && "coop-btn--loading",
    size === "small" && "coop-btn--small",
  ]);

  const classes = variant === "link" ? "coop-t-link" : dynamicClasses;

  return (
    <button {...props} className={classes}>
      {children}
    </button>
  );
};

Button.defaultProps = {
  type: "button",
  size: "medium",
  disabled: false,
  fullWidth: false,
  isLoading: false,
  className: null,
  variant: null,
  children: null,
};

Button.propTypes = {
  className: PropTypes.string,
  variant: PropTypes.oneOf(["primary", "white", "grey", "link"]),
  type: PropTypes.oneOf(["button", "submit", "reset"]),
  size: PropTypes.oneOf(["small", "medium"]),
  disabled: PropTypes.bool,
  fullWidth: PropTypes.bool,
  isLoading: PropTypes.bool,
  children: PropTypes.node,
};

export default Button;
