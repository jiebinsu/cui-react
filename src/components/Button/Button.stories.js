import * as React from "react";
import { action } from "@storybook/addon-actions";
import Button from "./Button";

export default {
  title: "Base/Button",
  component: Button,
};

export const Basic = (args) => <Button {...args} onClick={action("clicked")} />;
Basic.args = { variant: "primary", children: "Label" };

export const All = () => (
  <>
    <Button>Default</Button>
    <Button variant="primary">Primary</Button>
    <div style={{ background: "grey", display: "inline-block" }}>
      <Button variant="white">White</Button>
    </div>
    <Button variant="grey">Grey</Button>
    <Button size="small">Small</Button>
    <Button variant="link">Link</Button>
    <Button fullWidth>Full width</Button>
    <Button isLoading>Loading</Button>
  </>
);

export const Sizes = () => (
  <>
    <Button>Default</Button>
    <Button size="small">Small</Button>
  </>
);

export const Disabled = () => (
  <>
    <Button disabled>Primary</Button>
    <Button variant="primary" disabled>
      Primary
    </Button>
    <Button variant="white" disabled>
      White
    </Button>
    <Button variant="grey" disabled>
      Grey
    </Button>
  </>
);
