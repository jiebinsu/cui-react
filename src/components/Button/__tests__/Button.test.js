import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import Button from "../index";

describe("Button", () => {
  it("should render correctly", () => {
    const wrapper = mount(<Button>Button</Button>);
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it("should support all variants", () => {
    const Variants = () => (
      <div>
        <Button>Default</Button>
        <Button variant="primary">Primary</Button>
        <Button variant="white">White</Button>
        <Button variant="grey">Grey</Button>
        <Button variant="link">Link</Button>
      </div>
    );

    const tree = renderer.create(<Variants />).toJSON();
    expect(tree).toMatchSnapshot();

    const wrapper = mount(<Variants />);
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it("should support all sizes", () => {
    const Sizes = () => (
      <div>
        <Button size="small">Small</Button>
        <Button size="medium">Default/Medium</Button>
      </div>
    );

    const tree = renderer.create(<Sizes />).toJSON();
    expect(tree).toMatchSnapshot();

    const wrapper = mount(<Sizes />);
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it("should render different text", () => {
    const wrapper = mount(<Button>button</Button>);
    expect(wrapper.text()).toContain("button");

    wrapper.setProps({
      children: <span>New Text</span>,
    });
    expect(wrapper.text()).toContain("New Text");
  });

  it("should render empty button correctly", () => {
    const tree = renderer.create(<Button />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
