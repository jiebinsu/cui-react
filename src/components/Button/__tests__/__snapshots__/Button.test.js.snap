// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render empty button correctly 1`] = `
<button
  className="coop-btn"
/>
`;

exports[`Button should support all sizes 1`] = `
<div>
  <button
    className="coop-btn coop-btn--small"
  >
    Small
  </button>
  <button
    className="coop-btn"
  >
    Default/Medium
  </button>
</div>
`;

exports[`Button should support all variants 1`] = `
<div>
  <button
    className="coop-btn"
  >
    Default
  </button>
  <button
    className="coop-btn coop-btn--primary"
  >
    Primary
  </button>
  <button
    className="coop-btn coop-btn--white"
  >
    White
  </button>
  <button
    className="coop-btn coop-btn--grey"
  >
    Grey
  </button>
  <button
    className="coop-t-link"
  >
    Link
  </button>
</div>
`;
